test/tools/uri_spec_runner.ts

```diff
@@ -0,0 +1,304 @@
+  uri: string;
+  valid: boolean;
+};
+
+interface UriTest extends UriTestBase {
+  warning: boolean;
+  hosts: HostObject[];
+  auth: {
+    username: string;
+    password: string;
+    db: string;
+  };
+  options: Record<string, any>;
+}
+
+interface AuthTest extends UriTestBase {
+  credential: {
+    username: string;
+    password: string;
+    source: string;
+    mechanism: string;
+    mechanism_properties: Record<string, any>;
+  };
+}
+
+export function executeUriValidationTest(
+  test: UriTest | AuthTest,
+  shouldNotThrowOnWarn = false
+): void {
+  const knownTestKeys = [
+    'description',
+    'uri',
+    'valid',
+    'warning',
+    'hosts',
+    'auth',
+    'options',
+    'credential'
+  ];
+  expect(knownTestKeys).to.include.members(Object.keys(test));
+
+  const errorMessage = `"${test.uri}"`;
+
+  const valid = test.valid && (!(test as UriTest).warning || shouldNotThrowOnWarn);
+
+  if (!valid) {
+    try {
+      new MongoClient(test.uri);
+      expect.fail(`Expected "${test.uri}" to be invalid${test.valid ? ' because of warning' : ''}`);
+    } catch (err) {
+      if (err instanceof TypeError) {
+        expect(err).to.have.property('code').equal('ERR_INVALID_URL');
+      } else if (
+        // most of our validation is MongoParseError, which does not extend from MongoAPIError
+        !(err instanceof MongoParseError) &&
+        // the rest of our validation extends from MongoAPIError
+        !(err instanceof MongoAPIError) &&
+        // mongodb-connection-string-url does not export its MongoParseError so we can't check for it directly
+        err.name !== 'MongoParseError'
+      ) {
+        throw err;
+      }
+    }
+    return;
+  }
+
+  const client = new MongoClient(test.uri);
+  const options = client.options;
+  expect(options, errorMessage).to.be.an('object').that.is.not.empty;
+
+  if ('hosts' in test && test.hosts != null) {
+    for (const [index, { host, port }] of test.hosts.entries()) {
+      const actualHost = options.hosts[index];
+      if (actualHost.host == null && actualHost.socketPath == null) {
+        expect.fail('Expected host to define "host" or "socketPath" properties');
+      }
+      if (actualHost.host != null) {
+        expect(actualHost, errorMessage).property('host').to.equal(host);
+      } else {
+        expect(actualHost, errorMessage).property('socketPath').to.equal(host);
+      }
+
+      if (port != null) expect(actualHost).property('port').to.equal(port);
+    }
+  }
+
+  // depending on whether this is a UriTest or an AuthTest, credential option values are
+  // defined in test.credential or test.auth, respectively, and additional properties
+  // to test are defined in test.options or mixed into test.credential, respectively
+  let credentialsToTest: { source?: string; password?: string; username?: string } = {};
+  let optionsToTest: Record<string, any> = 'options' in test ? test.options || {} : {};
+
+  if (!('options' in test) && test.credential != null) {
```

Comment:
Is this being used to determine that the test is an AuthTest?  We could use a type assertion to clean it up

```typescript
function isAuthTest(test: UriTest | AuthTest) : test is AuthTest {
  return !('options' in test) && test.credential != null;
}

// then in the if statement
if (isAuthTest(test)) {
...
}
```

This might be overkill but I usually define them for all my discriminated unions so it's easier to understand and we get better type inference.

Reply:
yeah, it's a good idea to clean it up - as I mentioned in the other comment, I originally inlined the check for expediency when I didn't fully know what the types would end up being